### ISSUE TITLE
[feature request] np.packbits / np.unpackbits, general BitTensors (maybe can be just tensors with dtype torch.bits8 or have a new dtype torch.bits introduced) and bit packed tensors utilities for saving memory / accesses, support for BitTensors wherever BoolTensors are used

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/ao#292]
A usecase: storing a full backtracking pointer matrix can be okay for needleman/ctc alignment (4x memory saving compared to uint8 representation), if 2bit data type is used. Currently it's possible to do this with bit manipulation magic, but probably not very efficient (store and load will require masking and shifting, not fused)

Another usecase: compressed BoolTensor for binary neural networks

Another usecase: extremely low-bit quantized representations.

Is something like this already implemented for quantization? Probably a simple version of this feature could be providing some explicitly utility functions like calculating size of the holder `uint8` tensor, fused store and load functions (potentially explicitly batched, e.g. actual store is delayed until some aligned number of memory lines has arrived)

In NumPy the related functionality is `np.packbits` and `np.unpackbits`, however these are designed to work only with 1-bit contained type. 2-bit/4-bit would be cool as well.

On 1-bit side, another related project is RoaringBitmap https://github.com/RoaringBitmap/RoaringBitmap (http://roaringbitmap.org/) - for compressed bitsets for set operations.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @izdeby

**Comment by vadimkantorov** [comment]
I guess, for packbits/unpackbits ao is indeed a better place, but for more general BitTensor support in ops as compressed explicit-pack-calls-less alternative to BoolTensor - not sure if this discussion should stay in ao or back in core

**Comment by cpuhrsch** [comment]
@vadimkantorov - I think wherever the discussions happens (here or in core), it becomes easier once we have a prototype.

**Comment by vadimkantorov** [comment]
BitTensor would also be useful for compact storage/encoding of sign output (https://github.com/pytorch/pytorch/pull/130215) - or if TritTensor can be added to store on of three values: -1, 0, 1

**Comment by vadimkantorov** [comment]
https://github.com/microsoft/BitNet released code, so having some core support / ops support for torch.bit (and ternary tensors) might promote more research in extremely compressed binary nets

**Comment by vadimkantorov** [comment]
Curious, is now the compiled codegen for the most basic numpy-like packbots/unpackbits (bit-compression of bool tensors) comparable to handwritten kernels? If so - probably worth adding these to core torch

**Comment by vadimkantorov** [comment]
For BitNet representation (with values -1, 0, 1) (e.g. https://falcon-lm.github.io/blog/falcon-edge/), maybe a good name would be TritTensor :)

**Comment by vadimkantorov** [comment]
Regarding TritTensor:

More ternary kernels just released https://x.com/imtejas13/status/1951821672602099995

**Comment by vadimkantorov** [comment]
Related:
- https://github.com/pytorch/pytorch/issues/163821

**Comment by vadimkantorov** [comment]
Curious, why deemed not planned? It's very useful to be able to un/pack booltensor into a compact storage (and to manipulate compact masks) - also given that numpy has this capability
